src/System.Management.Automation/engine/InitialSessionState.cs

```diff
@@ -2360,67 +2360,104 @@ private void Bind_LoadAssemblies(ExecutionContext context)
-            HashSet<CommandInfo> publicCommands = new HashSet<CommandInfo>();
-            foreach (CommandInfo sessionCommand in initializedRunspace.ExecutionContext.SessionState.InvokeCommand.GetCommands(
-                        "*", CommandTypes.Alias | CommandTypes.Function | CommandTypes.Filter | CommandTypes.Cmdlet, true))
-            {
-                if (sessionCommand.Visibility == SessionStateEntryVisibility.Public)
+            // Get initial list of public commands from session in a lazy way.
+            // We could use Lazy<> with an initializer for the same purpose, but we can save allocations
+            // by using the local function. It avoids allocating the delegate, and it's more efficient on
+            // capturing variables from the enclosing scope by using a struct.
+            HashSet<CommandInfo> commands = null;
+            HashSet<CommandInfo> GetPublicCommands()
+            {
+                if (commands != null)
                 {
-                    publicCommands.Add(sessionCommand);
+                    return commands;
+                }
+
+                commands = new HashSet<CommandInfo>();
+                foreach (CommandInfo sessionCommand in initializedRunspace.ExecutionContext.SessionState.InvokeCommand.GetCommands(
+                            "*", CommandTypes.Alias | CommandTypes.Function | CommandTypes.Filter | CommandTypes.Cmdlet, nameIsPattern: true))
+                {
+                    if (sessionCommand.Visibility == SessionStateEntryVisibility.Public)
+                    {
+                        commands.Add(sessionCommand);
+                    }
                 }
+                return commands;
             }
 
-            // If a user has any module with the same name as that of the core module( or nested module inside the core module)
-            // in his module path, then that will get loaded instead of the actual nested module (from the GAC - in our case)
-            // Hence, searching only from the system module path while loading the core modules
             var unresolvedCmdsToExpose = new HashSet<string>(this.UnresolvedCommandsToExpose, StringComparer.OrdinalIgnoreCase);
-            ProcessImportModule(initializedRunspace, CoreModulesToImport, ModuleIntrinsics.GetPSHomeModulePath(), publicCommands, unresolvedCmdsToExpose);
+            if (CoreModulesToImport.Count > 0 || unresolvedCmdsToExpose.Count > 0)
+            {
+                // If a user has any module with the same name as that of the core module( or nested module inside the core module)
+                // in his module path, then that will get loaded instead of the actual nested module (from the GAC - in our case)
+                // Hence, searching only from the system module path while loading the core modules
+                ProcessImportModule(initializedRunspace, CoreModulesToImport, ModuleIntrinsics.GetPSHomeModulePath(), GetPublicCommands(), unresolvedCmdsToExpose);
```

Comment:
It looks like GetPublicCommands() result is only used in the `ProcessImportModule` overloaded method, so we don't need to pass it in as a parameter and can just add it there.  Also can we re-name this overload method?  I find it very confusing to have two methods that do different work named the same (and take the same number of parameter arguments).

Reply:
> It looks like GetPublicCommands() result is only used in the ProcessImportModule overloaded method, so we don't need to pass it in as a parameter and can just add it there

If we retrieve the public commands within `ProcessImportModule`, then in the worst case where we do have to call `ProcessImportModule` 3 times in `BindRunspace`, we will have to retrieve the public commands 3 time. That's redundant and should be avoided.

> Also can we re-name this overload method? I find it very confusing to have two methods that do different work named the same (and take the same number of parameter arguments).

I agree it's a bit confusing, especially one overload is only used in the other overload. I will try rename them.